Add MASTG Components Instructions (#3447)
OWASP/owasp-mstg: .github/instructions/mastg-test.instructions.md

## Tests

Tests are platform-specific and must be located under tests-beta/android/ or tests-beta/ios/, within the corresponding MASVS category. Their file names are the test IDs.

Notes:

> - Tests with `platform: network` are still organized under the OS folder that the MASVS category belongs to (for example, Android network tests live under `tests-beta/android/MASVS-NETWORK/`).

Reviewer: Readability-wise, it might be smoother to read after the `ls -l` "structure" example. Wider context->edge case

Tests have an overview and contain Steps which produce outputs called observations. After following the Steps you come up with an [Observation](https://docs.google.com/document/d/1EMsVdfrDBAu0gmjWAUEs60q-fWaOmDB5oecY9d9pOlg/edit?pli=1#heading=h.h9gqgz4hdubj) which you will [Evaluate](https://docs.google.com/document/d/1EMsVdfrDBAu0gmjWAUEs60q-fWaOmDB5oecY9d9pOlg/edit?pli=1#heading=h.lare0v58fwbf).

> Example:

Reviewer: Maybe underline the reason why we have this example:

```suggestion
Example structure:
```

```sh
% ls -1 -F tests-beta/android/MASVS-CRYPTO/
MASTG-TEST-0204.md
MASTG-TEST-0205.md
```

> 

Reviewer: Add structural explanation and links to sessions:

```suggestion
Each test has two parts: the [Markdown metadata](#markdown-metadata) (YAML `front matter`) and the [Markdown body](markdown-body).

```

### Markdown: Metadata

#### title

Test titles should be concise and clearly state the purpose of the test.

In some cases, the test name and the weakness may have the same title, but typically tests cover different aspects of a weakness. Titles should reflect that.

Avoid including Android or iOS unless necessary, as in "Insecure use of the Android Protected Confirmation API".

Follow a consistent style across all test titles.

**Conventions**

- Static: “References to…” (semgrep/r2)
- Dynamic: “Runtime Use …” (frida)

Exceptions may apply where "Runtime ..." feels forced, for example tests using adb, local backups, or filesystem snapshots.

Examples:

- [MASTG-TEST-0207](https://mas.owasp.org/MASTG/tests-beta/android/MASVS-STORAGE/MASTG-TEST-0207/)
- [MASTG-TEST-0216](https://mas.owasp.org/MASTG/tests-beta/android/MASVS-STORAGE/MASTG-TEST-0216/)

> - [MASTG-TEST-0263](https://mas.owasp.org/MASTG/tests-beta/android/MASVS-STORAGE/MASTG-TEST-0263/)

Reviewer: Maybe move them to the `## Tests` session as they are not "title-specific" and explain the reason i.e. `Tests to use as reference`

#### platform

The mobile platform. One of: ios, android, network.

- Use network for platform-agnostic traffic analysis tests where the checks are performed purely on captured/observed traffic (often paired with type: [network]).

#### id

The test ID.

#### weakness

The MASWE weakness ID the test references.

- In YAML front matter, specify the bare identifier (for example, `weakness: MASWE-0069`). In body text, include the leading `@` (for example, @MASWE-0069).

#### type

One or more test types.

Supported: static, dynamic, network, manual, filesystem, developer.

> 

Reviewer: Static & dynamic are self-explanatory but could be defined. The rest are not very obvious, so a small description might help the reader. Sample wording:

```suggestion

static: performed without executing the application (e.g by reading the source code)
dynamic: performed by executing the binary and testing it within a (closed) runtime environment
network: TODO
manual: tests that you have to use the app
filesystem: TODO
```

- developer: tests that work on developer artifacts or supply-chain outputs (for example SBOM, dependency manager lockfiles, scanner reports) rather than the built app.

Example:

```md
type: [static]
```

Examples with multiple types:

```md
type: [dynamic, manual]
```

```md
type: [network]
```

#### best-practices

Reference platform-specific mitigations or best practices. Automation generates a “Mitigations” section.

New best practice files can be added under [best-practices/](https://github.com/OWASP/owasp-mastg/tree/master/best-practices).

Example:

```md
best-practices: [MASTG-BEST-0001]
```

This links to https://mas.owasp.org/MASTG/best-practices/MASTG-BEST-0001/

Notes:

- If no applicable best practices exist yet, you can omit the field or set an empty list: `best-practices: []`.

#### prerequisites

> List prerequisites needed to execute or evaluate the test. Existing files are in prerequisites/. Create new ones if needed.

Reviewer: ```suggestion
List prerequisites needed to execute or evaluate the test. Existing files are in `prerequisites/`. Create new ones if needed.
```

- If there are no prerequisites, you can omit this field or use an empty list.

Example:

```md
prerequisites:
- identify-sensitive-data
- identify-security-relevant-contexts
```

#### profiles

Specify MASVS profiles where the test applies. Valid values: L1, L2, P, R.

> 

Reviewer: ```suggestion
The profiles are described in [MAS Testing Profiles Guide]( https://docs.google.com/document/d/1paz7dxKXHzAC9MN7Mnln1JiZwBNyg7Gs364AJ6KudEs/edit?tab=t.0#heading=h.il6q80u4fm3n)

L1 denotes Essential Security
L2 denotes Advanced Security
```

- P denotes Privacy.
- R denotes Resilience.

Example:

```md
profiles: [L1, L2, P]
```

#### optional fields

Include these if relevant:

- `status:` draft, placeholder, deprecated
- `note:` short free-form note
- `available_since:` minimum platform/API level
- `deprecated_since:` last applicable platform/API level

> 

Reviewer: ```suggestion
`apis: TODO` 
```

#### id

The test ID.

#### weakness

The MASWE weakness ID associated with this test.

- In YAML front matter, specify the bare identifier (for example, `weakness: MASWE-0069`). In body text, include the leading `@` (for example, @MASWE-0069).

#### type

One or more test types.

Supported:

- `static`: analysis of the app binary, reverse-engineered source code, or developer artifacts that are available in the APK/IPA app package (e.g., Android manifest, Info.plist, entitlements, etc.). No execution of the app is required.
- `dynamic`: analysis of the app while it is running and involves runtime analysis such as hooking or method tracing.
- `manual`: manual steps that require human judgment, such as inspecting app behavior, UI, or configuration. This may include reverse engineering or runtime analysis that cannot be fully automated.
- `network`: analysis of network traffic, while the app is running. Done externally, for example, using a proxy or network capture tool.
- `filesystem`: analysis of the app's file system, including local storage or backups, which doesn't involve runtime analysis such as hooking or method tracing.

> - `developer`: tests only the developer can perform because they require access to the source code, build process, or other internal resources.

Author: Suggestion by @TheDauntless 

```suggestion
- `source-code`: tests only the developer can perform because they require access to the source code, build process, or other internal resources.
```

Example:

```md
type: [static]
```

Examples with multiple types:

```md
type: [dynamic, manual]
```

#### best-practices

Reference platform-specific mitigations or best practices. Automation generates a “Mitigations” section.

New best practice files can be added under [best-practices/](https://github.com/OWASP/owasp-mastg/tree/master/best-practices).

Example:

```md
best-practices: [MASTG-BEST-0001]
```

This links to https://mas.owasp.org/MASTG/best-practices/MASTG-BEST-0001/

Notes:

- If no applicable best practices exist yet, you can omit the field or set an empty list: `best-practices: []`.

#### prerequisites

> List the prerequisites needed to execute or evaluate the test. Existing files are in `prerequisites/`. Create new ones if required.

Author: ```suggestion
List the conditions that must be known or available before running or evaluating the test. These items capture internal context that only the developer or the organization can provide. Existing files are in the `prerequisites/` folder. Create new ones when needed.

Common examples include:

- Defined categories of sensitive data and their sensitivity levels used within the app.
- A list of first party packages, libraries, and modules.
- A list of first party network domains and services the app is expected to contact.
```

> - If there are no prerequisites, you can omit this field or use an empty list.

Author: ```suggestion
If there are no prerequisites, you can omit this field or use an empty list.
```
